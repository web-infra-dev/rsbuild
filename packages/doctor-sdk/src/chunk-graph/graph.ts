import { SDK } from '@rsbuild/doctor-types';
import type { Asset } from './asset';
import type { Chunk } from './chunk';
import type { Module } from '../module-graph';
import type { EntryPoint } from './entrypoint';

export class ChunkGraph implements SDK.ChunkGraphInstance {
  private _assets: Asset[] = [];

  private _chunks: Chunk[] = [];

  private _entrypoints: EntryPoint[] = [];

  getAssets(): Asset[] {
    return this._assets.slice();
  }

  getChunks(): Chunk[] {
    return this._chunks.slice();
  }

  addAsset(...assets: Asset[]): void {
    assets.forEach((asset) => {
      if (!this._assets.includes(asset)) {
        this._assets.push(asset);
      }
    });
  }

  addChunk(...chunks: Chunk[]): void {
    chunks.forEach((chunk) => {
      if (!this._chunks.includes(chunk)) {
        this._chunks.push(chunk);
      }
    });
  }

  getChunkById(id: string): Chunk | undefined {
    return this._chunks.find((item) => item.id === id);
  }

  getChunkByModule(module: Module): Chunk | undefined {
    return this._chunks.find((item) => item.hasModule(module));
  }

  getAssetByPath(path: string): Asset | undefined {
    return this._assets.find((item) => item.path === path);
  }

  getAssetsByChunk(chunk: Chunk): Asset[] | undefined {
    return this._assets.filter((item) => {
      const _chunk = item.chunks.find((ck) => ck.id === chunk.id);
      if (_chunk) return true;
    });
  }

  getEntryPoints(): EntryPoint[] {
    return this._entrypoints.slice();
  }

  addEntryPoint(...entrypoints: EntryPoint[]): void {
    entrypoints.forEach((entrypoint) => {
      if (!this._entrypoints.includes(entrypoint)) {
        this._entrypoints.push(entrypoint);
      }
    });
  }

  /** output the chunk graph data */
  toData(type: SDK.ToDataType): SDK.ChunkGraphData {
    return {
      assets: this._assets.map((item) => item.toData(type)),
      chunks: this._chunks.map((item) => item.toData()),
      entrypoints: this._entrypoints.map((item) => item.toData()),
    };
  }
}
