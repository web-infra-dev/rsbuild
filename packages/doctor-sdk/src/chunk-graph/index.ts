export * from './chunk';
export * from './graph';
export * from './asset';
export * from './entrypoint';
