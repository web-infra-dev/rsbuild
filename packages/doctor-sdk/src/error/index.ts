export * from './error';

export { printErrors } from './utils';
