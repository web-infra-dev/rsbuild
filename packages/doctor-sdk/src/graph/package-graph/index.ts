export * from './dependency';
export * from './graph';
export * from './package';
export * from './types';
