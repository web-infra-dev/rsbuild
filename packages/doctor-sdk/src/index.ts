export * from './chunk-graph';
export * from './module-graph';
export * from './package-graph';
