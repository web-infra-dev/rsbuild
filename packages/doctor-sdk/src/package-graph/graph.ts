import { unionBy } from 'lodash';
import { dirname, join, resolve } from 'path';
import { SDK } from '@rsbuild/doctor-types';
import { Package as PackageUtil } from '@rsbuild/doctor-utils/common';
import type { ModuleGraph, Module } from '../module-graph';
import { Package } from './package';
import { PackageDependency } from './dependency';

export class PackageGraph implements SDK.PackageGraphInstance {
  static fromModuleGraph(
    graph: ModuleGraph,
    root: string,
    getPackageFile?: SDK.GetPackageFile,
  ): PackageGraph {
    const pkgGraph = new PackageGraph(root);
    const modules = graph
      .getModules()
      .filter((item) => item.kind === SDK.ModuleKind.Normal);

    // Generate all package data.
    for (const item of modules) {
      const pkg = pkgGraph.getPackageByModule(item, getPackageFile);

      if (pkg) {
        pkgGraph.addPackage(pkg);
        pkg.addModule(item);
      }
    }

    // Generate all dependent data.
    for (const dep of graph.getDependencies()) {
      const modulePkg = pkgGraph.getPackageByFile(dep.module.path);
      const dependencyPkg = pkgGraph.getPackageByFile(dep.dependency.path);

      if (modulePkg && dependencyPkg && !modulePkg.isSame(dependencyPkg)) {
        const pkgDep = new PackageDependency(modulePkg, dependencyPkg, dep);
        pkgGraph.addDependency(pkgDep);
        modulePkg.addDependency(pkgDep);
      }
    }

    return pkgGraph;
  }

  private _root: string;

  private _dependencies: PackageDependency[] = [];

  private _packages: Package[] = [];

  private _pkgNameMap = new Map<string, Package[]>();

  private _pkgFileMap = new Map<string, Package>();

  constructor(root: string) {
    this._root = root;
  }

  getPackages() {
    return this._packages.slice();
  }

  getPackageByModule(module: Module, readFile?: SDK.GetPackageFile) {
    const { path: file, meta } = module;
    const { _pkgFileMap: pkgsMap } = this;
    const getPackageByData = (data: SDK.PackageBasicData) => {
      return (
        this.getPackageByData(data) ??
        new Package(data.name, data.root, data.version)
      );
    };

    if (pkgsMap.has(file)) {
      return pkgsMap.get(file);
    }

    if (meta.packageData) {
      const pkg = getPackageByData(meta.packageData);
      pkgsMap.set(file, pkg);
      return pkg;
    }

    const readPackageJson = (
      file: string,
      readFile?: SDK.GetPackageFile,
    ): SDK.PackageBasicData | undefined => {
      let result: SDK.PackageJSONData | undefined;
      let current = file;

      while (current !== '/' && !result) {
        if (dirname(current) === current) {
          break;
        }
        current = dirname(current);
        if (readFile) {
          result =
            readFile(join(current, 'package.json')) ||
            PackageUtil.getPackageMetaFromModulePath(file);
        }
        if (!readFile || !result?.name) {
          result = PackageUtil.getPackageMetaFromModulePath(file);
        }
      }

      if (!result) {
        return;
      }

      // Some packages will put an empty package.json in the source folder.
      if (readFile && (!result.name || !result.version)) {
        return readPackageJson(dirname(current), readFile);
      }

      return {
        ...result,
        root: current,
      };
    };

    const cache = this.getPackageContainFile(file);

    if (cache) {
      pkgsMap.set(file, cache);
      return cache;
    }
    const data = readPackageJson(file, readFile);

    if (!data) {
      return;
    }

    if (data.root.startsWith('.')) {
      data.root = resolve(this._root, data.root);
    }

    const pkg = getPackageByData(data);

    this.addPackage(pkg);
    pkgsMap.set(file, pkg);
    return pkg;
  }

  getPackageByFile(file: string) {
    return this._pkgFileMap.get(file);
  }

  getPackageContainFile(file: string) {
    return this._packages.find((pkg) => pkg.contain(file));
  }

  getPackagesByName(name: string) {
    return this._pkgNameMap.get(name) ?? [];
  }

  getPackageByData(data: SDK.PackageBasicData) {
    return this._pkgNameMap
      .get(data.name)
      ?.find(
        (item) => item.version === data.version && item.root === data.root,
      );
  }

  addPackage(pkg: Package) {
    if (this._packages.every((item) => !item.isSame(pkg))) {
      this._packages.push(pkg);

      const { _pkgNameMap: map } = this;
      const arr = map.get(pkg.name) ?? [];

      if (arr.every((item) => !item.isSame(pkg))) {
        arr.push(pkg);
        map.set(pkg.name, arr);
      }
    }
  }

  getDependenciesFromPackage(pkg: Package) {
    return this._dependencies.filter((dep) => dep.dependency === pkg);
  }

  addDependency(dep: PackageDependency) {
    if (this._dependencies.every((item) => !item.isSame(dep))) {
      this._dependencies.push(dep);
    }
  }

  getDependenciesFromOrigin() {
    return this._dependencies.filter((item) => !item.package);
  }

  getDuplicatePackages(): Package[][] {
    return unionBy(
      Array.from(this._pkgNameMap.values()).filter((pkgs) => pkgs.length > 1),
      (pkgs) => pkgs[0].name,
    );
  }

  toData(): SDK.PackageGraphData {
    return {
      packages: this._packages.map((e) => e.toData()),
      dependencies: this._dependencies.map((d) => d.toData()),
    };
  }
}
