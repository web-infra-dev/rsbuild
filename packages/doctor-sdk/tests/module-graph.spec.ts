import path from 'path';
import { expect, describe, it } from 'vitest';
import type { SDK } from '@rsbuild/doctor-types';
import { Module, ModuleGraph, PackageGraph } from '../src/graph';

const resolveFixture = (...paths: string[]) => {
  return path.resolve(__dirname, 'fixture', ...paths);
};

function arrayEq<T>(actual: T[], expected: T[]) {
  expect(actual.length).toBe(expected.length);
  expect(actual).toEqual(expect.arrayContaining(expected));
}

describe('module graph', () => {
  it('from space data', async () => {
    expect(ModuleGraph.fromData({}).toData()).toStrictEqual({
      modules: [],
      dependencies: [],
      exports: [],
      moduleGraphModules: [],
      sideEffects: [],
      variables: [],
    });
  });

  it('from data basic', async () => {
    const inputData = require(
      resolveFixture('module-graph-basic.json'),
    ) as SDK.ModuleGraphData;
    const moduleGraph = ModuleGraph.fromData(inputData);

    expect(moduleGraph.size()).toBe(inputData.modules.length);

    inputData.modules.forEach((item) => {
      const module = moduleGraph.getModuleById(item.id)!;
      expect(module).toBeTruthy();
      expect(module.kind).toBe(item.kind);
      expect(module.getSize()).toStrictEqual(item.size);
      arrayEq(
        module.getDependencies().map((item) => item.id),
        item.dependencies,
      );
      arrayEq(
        module.getImported().map((item) => item.id),
        item.imported,
      );

      if (!item.meta) {
        expect(module.meta).toStrictEqual({
          hasSetEsModuleStatement: false,
          strictHarmonyModule: false,
        });
      } else {
        expect(module.meta.hasSetEsModuleStatement).toStrictEqual(
          item.meta.hasSetEsModuleStatement ?? false,
        );
        expect(module.meta.strictHarmonyModule).toStrictEqual(
          item.meta.strictHarmonyModule ?? false,
        );
      }
    });

    inputData.dependencies.forEach((item) => {
      const dependency = moduleGraph.getDependencyById(item.id)!;
      expect(dependency).toBeTruthy();
      expect(dependency.id).toBe(item.id);
      expect(dependency.request).toBe(item.request);
      expect(dependency.resolvedRequest).toBe(item.resolvedRequest);
      expect(dependency.module.id).toBe(item.module);
      expect(dependency.dependency.id).toBe(item.dependency);
    });

    const pkgGraph = PackageGraph.fromModuleGraph(moduleGraph, '.');
    const pkgData = pkgGraph.toData();
    expect(pkgData.packages[0].root).toBeTruthy(); // TODO: test error
    pkgData.packages.forEach((pkg) => (pkg.root = ''));
    expect(pkgData).toMatchSnapshot();
  });

  it('ModuleGraph toCodeMap', async () => {
    const _moduleGraph = new ModuleGraph();
    for (let i = 1; i < 10; i++) {
      const _mod = new Module(i.toString(), `index-${i}.js`);
      _mod.setSource({
        source: `source-${i}`,
        transformed: `source-${i}`,
        parsedSource: `source-${i}`,
      });
      _moduleGraph.addModule(_mod);
    }

    expect(_moduleGraph.toCodeData()).toMatchSnapshot();
  });
});
