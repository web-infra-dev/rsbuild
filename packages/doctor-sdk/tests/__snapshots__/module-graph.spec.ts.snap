// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`module graph > ModuleGraph toCodeMap 1`] = `
{
  "1": {
    "parsedSource": "source-1",
    "source": "source-1",
    "transformed": "source-1",
  },
  "2": {
    "parsedSource": "source-2",
    "source": "source-2",
    "transformed": "source-2",
  },
  "3": {
    "parsedSource": "source-3",
    "source": "source-3",
    "transformed": "source-3",
  },
  "4": {
    "parsedSource": "source-4",
    "source": "source-4",
    "transformed": "source-4",
  },
  "5": {
    "parsedSource": "source-5",
    "source": "source-5",
    "transformed": "source-5",
  },
  "6": {
    "parsedSource": "source-6",
    "source": "source-6",
    "transformed": "source-6",
  },
  "7": {
    "parsedSource": "source-7",
    "source": "source-7",
    "transformed": "source-7",
  },
  "8": {
    "parsedSource": "source-8",
    "source": "source-8",
    "transformed": "source-8",
  },
  "9": {
    "parsedSource": "source-9",
    "source": "source-9",
    "transformed": "source-9",
  },
}
`;

exports[`module graph > from data basic 1`] = `
{
  "dependencies": [
    {
      "dependency": 2,
      "id": 1,
      "package": 1,
      "refDependency": 84,
    },
    {
      "dependency": 5,
      "id": 2,
      "package": 1,
      "refDependency": 85,
    },
    {
      "dependency": 6,
      "id": 3,
      "package": 1,
      "refDependency": 86,
    },
    {
      "dependency": 2,
      "id": 4,
      "package": 1,
      "refDependency": 10,
    },
    {
      "dependency": 5,
      "id": 5,
      "package": 1,
      "refDependency": 15,
    },
    {
      "dependency": 6,
      "id": 6,
      "package": 1,
      "refDependency": 16,
    },
    {
      "dependency": 3,
      "id": 7,
      "package": 2,
      "refDependency": 11,
    },
    {
      "dependency": 4,
      "id": 8,
      "package": 2,
      "refDependency": 12,
    },
    {
      "dependency": 7,
      "id": 9,
      "package": 3,
      "refDependency": 17,
    },
    {
      "dependency": 8,
      "id": 10,
      "package": 5,
      "refDependency": 18,
    },
    {
      "dependency": 9,
      "id": 11,
      "package": 5,
      "refDependency": 19,
    },
    {
      "dependency": 10,
      "id": 12,
      "package": 5,
      "refDependency": 20,
    },
    {
      "dependency": 11,
      "id": 13,
      "package": 6,
      "refDependency": 32,
    },
    {
      "dependency": 12,
      "id": 14,
      "package": 6,
      "refDependency": 37,
    },
    {
      "dependency": 13,
      "id": 15,
      "package": 6,
      "refDependency": 40,
    },
    {
      "dependency": 14,
      "id": 16,
      "package": 10,
      "refDependency": 46,
    },
    {
      "dependency": 15,
      "id": 17,
      "package": 10,
      "refDependency": 47,
    },
    {
      "dependency": 16,
      "id": 18,
      "package": 10,
      "refDependency": 48,
    },
    {
      "dependency": 17,
      "id": 19,
      "package": 6,
      "refDependency": 51,
    },
    {
      "dependency": 17,
      "id": 20,
      "package": 6,
      "refDependency": 52,
    },
    {
      "dependency": 12,
      "id": 21,
      "package": 11,
      "refDependency": 34,
    },
    {
      "dependency": 11,
      "id": 22,
      "package": 6,
      "refDependency": 31,
    },
    {
      "dependency": 13,
      "id": 23,
      "package": 6,
      "refDependency": 39,
    },
    {
      "dependency": 11,
      "id": 24,
      "package": 13,
      "refDependency": 26,
    },
    {
      "dependency": 18,
      "id": 25,
      "package": 7,
      "refDependency": 64,
    },
    {
      "dependency": 19,
      "id": 26,
      "package": 15,
      "refDependency": 65,
    },
    {
      "dependency": 12,
      "id": 27,
      "package": 11,
      "refDependency": 35,
    },
    {
      "dependency": 11,
      "id": 28,
      "package": 13,
      "refDependency": 29,
    },
    {
      "dependency": 12,
      "id": 29,
      "package": 13,
      "refDependency": 36,
    },
    {
      "dependency": 20,
      "id": 30,
      "package": 13,
      "refDependency": 68,
    },
    {
      "dependency": 11,
      "id": 31,
      "package": 13,
      "refDependency": 30,
    },
    {
      "dependency": 11,
      "id": 32,
      "package": 13,
      "refDependency": 28,
    },
    {
      "dependency": 11,
      "id": 33,
      "package": 13,
      "refDependency": 27,
    },
    {
      "dependency": 11,
      "id": 34,
      "package": 13,
      "refDependency": 25,
    },
    {
      "dependency": 12,
      "id": 35,
      "package": 20,
      "refDependency": 33,
    },
    {
      "dependency": 21,
      "id": 36,
      "package": 20,
      "refDependency": 72,
    },
    {
      "dependency": 22,
      "id": 37,
      "package": 19,
      "refDependency": 74,
    },
  ],
  "packages": [
    {
      "id": 1,
      "modules": [
        2,
        3,
        4,
        5,
        6,
        7,
        8,
      ],
      "name": "",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 841,
        "transformedSize": 0,
      },
      "version": "",
    },
    {
      "id": 2,
      "modules": [
        9,
        12,
        13,
      ],
      "name": "chalk",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 10477,
        "transformedSize": 0,
      },
      "version": "4.1.2",
    },
    {
      "id": 3,
      "modules": [
        10,
      ],
      "name": "ansi-styles",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 4139,
        "transformedSize": 0,
      },
      "version": "4.3.0",
    },
    {
      "id": 4,
      "modules": [
        11,
      ],
      "name": "supports-color",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 67,
        "transformedSize": 0,
      },
      "version": "7.2.0",
    },
    {
      "id": 5,
      "modules": [
        14,
      ],
      "name": "@babel/highlight",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 2719,
        "transformedSize": 0,
      },
      "version": "",
    },
    {
      "id": 6,
      "modules": [
        15,
        20,
        21,
        24,
      ],
      "name": "htmlparser2",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 53921,
        "transformedSize": 0,
      },
      "version": "7.2.0",
    },
    {
      "id": 7,
      "modules": [
        16,
        26,
        27,
      ],
      "name": "color-convert",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 21005,
        "transformedSize": 0,
      },
      "version": "2.0.1",
    },
    {
      "id": 8,
      "modules": [
        17,
      ],
      "name": "js-tokens",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 1448,
        "transformedSize": 0,
      },
      "version": "4.0.0",
    },
    {
      "id": 9,
      "modules": [
        18,
        28,
        29,
      ],
      "name": "@babel/helper-validator-identifier",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 15147,
        "transformedSize": 0,
      },
      "version": "",
    },
    {
      "id": 10,
      "modules": [
        19,
        33,
      ],
      "name": "chalk",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 9572,
        "transformedSize": 0,
      },
      "version": "2.4.2",
    },
    {
      "id": 11,
      "modules": [
        22,
        36,
      ],
      "name": "domhandler",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 21324,
        "transformedSize": 0,
      },
      "version": "4.3.1",
    },
    {
      "id": 12,
      "modules": [
        23,
      ],
      "name": "domelementtype",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 1978,
        "transformedSize": 0,
      },
      "version": "2.3.0",
    },
    {
      "id": 13,
      "modules": [
        25,
        37,
        38,
        39,
        40,
        41,
        42,
        43,
      ],
      "name": "domutils",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 31488,
        "transformedSize": 0,
      },
      "version": "2.8.0",
    },
    {
      "id": 14,
      "modules": [
        30,
      ],
      "name": "escape-string-regexp",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 226,
        "transformedSize": 0,
      },
      "version": "1.0.5",
    },
    {
      "id": 15,
      "modules": [
        31,
      ],
      "name": "ansi-styles",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 3574,
        "transformedSize": 0,
      },
      "version": "3.2.1",
    },
    {
      "id": 16,
      "modules": [
        32,
      ],
      "name": "supports-color",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 67,
        "transformedSize": 0,
      },
      "version": "5.5.0",
    },
    {
      "id": 17,
      "modules": [
        34,
        35,
        46,
        47,
      ],
      "name": "entities",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 106903,
        "transformedSize": 0,
      },
      "version": "3.0.1",
    },
    {
      "id": 18,
      "modules": [
        44,
      ],
      "name": "color-name",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 4617,
        "transformedSize": 0,
      },
      "version": "1.1.4",
    },
    {
      "id": 19,
      "modules": [
        45,
        49,
        50,
      ],
      "name": "color-convert",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 20802,
        "transformedSize": 0,
      },
      "version": "1.9.3",
    },
    {
      "id": 20,
      "modules": [
        48,
        52,
      ],
      "name": "dom-serializer",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 10115,
        "transformedSize": 0,
      },
      "version": "1.4.1",
    },
    {
      "id": 21,
      "modules": [
        51,
        54,
        55,
        56,
        57,
        58,
        59,
        60,
      ],
      "name": "entities",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 46808,
        "transformedSize": 0,
      },
      "version": "2.2.0",
    },
    {
      "id": 22,
      "modules": [
        53,
      ],
      "name": "color-name",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 4617,
        "transformedSize": 0,
      },
      "version": "1.1.3",
    },
  ],
}
`;

exports[`module graph ModuleGraph toCodeMap 1`] = `
{
  "1": {
    "parsedSource": "source-1",
    "source": "source-1",
    "transformed": "source-1",
  },
  "2": {
    "parsedSource": "source-2",
    "source": "source-2",
    "transformed": "source-2",
  },
  "3": {
    "parsedSource": "source-3",
    "source": "source-3",
    "transformed": "source-3",
  },
  "4": {
    "parsedSource": "source-4",
    "source": "source-4",
    "transformed": "source-4",
  },
  "5": {
    "parsedSource": "source-5",
    "source": "source-5",
    "transformed": "source-5",
  },
  "6": {
    "parsedSource": "source-6",
    "source": "source-6",
    "transformed": "source-6",
  },
  "7": {
    "parsedSource": "source-7",
    "source": "source-7",
    "transformed": "source-7",
  },
  "8": {
    "parsedSource": "source-8",
    "source": "source-8",
    "transformed": "source-8",
  },
  "9": {
    "parsedSource": "source-9",
    "source": "source-9",
    "transformed": "source-9",
  },
}
`;

exports[`module graph from data basic 1`] = `
{
  "dependencies": [
    {
      "dependency": 2,
      "id": 1,
      "package": 1,
      "refDependency": 84,
    },
    {
      "dependency": 5,
      "id": 2,
      "package": 1,
      "refDependency": 85,
    },
    {
      "dependency": 6,
      "id": 3,
      "package": 1,
      "refDependency": 86,
    },
    {
      "dependency": 2,
      "id": 4,
      "package": 1,
      "refDependency": 10,
    },
    {
      "dependency": 5,
      "id": 5,
      "package": 1,
      "refDependency": 15,
    },
    {
      "dependency": 6,
      "id": 6,
      "package": 1,
      "refDependency": 16,
    },
    {
      "dependency": 3,
      "id": 7,
      "package": 2,
      "refDependency": 11,
    },
    {
      "dependency": 4,
      "id": 8,
      "package": 2,
      "refDependency": 12,
    },
    {
      "dependency": 7,
      "id": 9,
      "package": 3,
      "refDependency": 17,
    },
    {
      "dependency": 8,
      "id": 10,
      "package": 5,
      "refDependency": 18,
    },
    {
      "dependency": 9,
      "id": 11,
      "package": 5,
      "refDependency": 19,
    },
    {
      "dependency": 10,
      "id": 12,
      "package": 5,
      "refDependency": 20,
    },
    {
      "dependency": 11,
      "id": 13,
      "package": 6,
      "refDependency": 32,
    },
    {
      "dependency": 12,
      "id": 14,
      "package": 6,
      "refDependency": 37,
    },
    {
      "dependency": 13,
      "id": 15,
      "package": 6,
      "refDependency": 40,
    },
    {
      "dependency": 14,
      "id": 16,
      "package": 10,
      "refDependency": 46,
    },
    {
      "dependency": 15,
      "id": 17,
      "package": 10,
      "refDependency": 47,
    },
    {
      "dependency": 16,
      "id": 18,
      "package": 10,
      "refDependency": 48,
    },
    {
      "dependency": 17,
      "id": 19,
      "package": 6,
      "refDependency": 51,
    },
    {
      "dependency": 17,
      "id": 20,
      "package": 6,
      "refDependency": 52,
    },
    {
      "dependency": 12,
      "id": 21,
      "package": 11,
      "refDependency": 34,
    },
    {
      "dependency": 11,
      "id": 22,
      "package": 6,
      "refDependency": 31,
    },
    {
      "dependency": 13,
      "id": 23,
      "package": 6,
      "refDependency": 39,
    },
    {
      "dependency": 11,
      "id": 24,
      "package": 13,
      "refDependency": 26,
    },
    {
      "dependency": 18,
      "id": 25,
      "package": 7,
      "refDependency": 64,
    },
    {
      "dependency": 19,
      "id": 26,
      "package": 15,
      "refDependency": 65,
    },
    {
      "dependency": 12,
      "id": 27,
      "package": 11,
      "refDependency": 35,
    },
    {
      "dependency": 11,
      "id": 28,
      "package": 13,
      "refDependency": 29,
    },
    {
      "dependency": 12,
      "id": 29,
      "package": 13,
      "refDependency": 36,
    },
    {
      "dependency": 20,
      "id": 30,
      "package": 13,
      "refDependency": 68,
    },
    {
      "dependency": 11,
      "id": 31,
      "package": 13,
      "refDependency": 30,
    },
    {
      "dependency": 11,
      "id": 32,
      "package": 13,
      "refDependency": 28,
    },
    {
      "dependency": 11,
      "id": 33,
      "package": 13,
      "refDependency": 27,
    },
    {
      "dependency": 11,
      "id": 34,
      "package": 13,
      "refDependency": 25,
    },
    {
      "dependency": 12,
      "id": 35,
      "package": 20,
      "refDependency": 33,
    },
    {
      "dependency": 21,
      "id": 36,
      "package": 20,
      "refDependency": 72,
    },
    {
      "dependency": 22,
      "id": 37,
      "package": 19,
      "refDependency": 74,
    },
  ],
  "packages": [
    {
      "id": 1,
      "modules": [
        2,
        3,
        4,
        5,
        6,
        7,
        8,
      ],
      "name": "",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 841,
        "transformedSize": 0,
      },
      "version": "",
    },
    {
      "id": 2,
      "modules": [
        9,
        12,
        13,
      ],
      "name": "chalk",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 10477,
        "transformedSize": 0,
      },
      "version": "4.1.2",
    },
    {
      "id": 3,
      "modules": [
        10,
      ],
      "name": "ansi-styles",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 4139,
        "transformedSize": 0,
      },
      "version": "4.3.0",
    },
    {
      "id": 4,
      "modules": [
        11,
      ],
      "name": "supports-color",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 67,
        "transformedSize": 0,
      },
      "version": "7.2.0",
    },
    {
      "id": 5,
      "modules": [
        14,
      ],
      "name": "@babel/highlight",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 2719,
        "transformedSize": 0,
      },
      "version": "",
    },
    {
      "id": 6,
      "modules": [
        15,
        20,
        21,
        24,
      ],
      "name": "htmlparser2",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 53921,
        "transformedSize": 0,
      },
      "version": "7.2.0",
    },
    {
      "id": 7,
      "modules": [
        16,
        26,
        27,
      ],
      "name": "color-convert",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 21005,
        "transformedSize": 0,
      },
      "version": "2.0.1",
    },
    {
      "id": 8,
      "modules": [
        17,
      ],
      "name": "js-tokens",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 1448,
        "transformedSize": 0,
      },
      "version": "4.0.0",
    },
    {
      "id": 9,
      "modules": [
        18,
        28,
        29,
      ],
      "name": "@babel/helper-validator-identifier",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 15147,
        "transformedSize": 0,
      },
      "version": "",
    },
    {
      "id": 10,
      "modules": [
        19,
        33,
      ],
      "name": "chalk",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 9572,
        "transformedSize": 0,
      },
      "version": "2.4.2",
    },
    {
      "id": 11,
      "modules": [
        22,
        36,
      ],
      "name": "domhandler",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 21324,
        "transformedSize": 0,
      },
      "version": "4.3.1",
    },
    {
      "id": 12,
      "modules": [
        23,
      ],
      "name": "domelementtype",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 1978,
        "transformedSize": 0,
      },
      "version": "2.3.0",
    },
    {
      "id": 13,
      "modules": [
        25,
        37,
        38,
        39,
        40,
        41,
        42,
        43,
      ],
      "name": "domutils",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 31488,
        "transformedSize": 0,
      },
      "version": "2.8.0",
    },
    {
      "id": 14,
      "modules": [
        30,
      ],
      "name": "escape-string-regexp",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 226,
        "transformedSize": 0,
      },
      "version": "1.0.5",
    },
    {
      "id": 15,
      "modules": [
        31,
      ],
      "name": "ansi-styles",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 3574,
        "transformedSize": 0,
      },
      "version": "3.2.1",
    },
    {
      "id": 16,
      "modules": [
        32,
      ],
      "name": "supports-color",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 67,
        "transformedSize": 0,
      },
      "version": "5.5.0",
    },
    {
      "id": 17,
      "modules": [
        34,
        35,
        46,
        47,
      ],
      "name": "entities",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 106903,
        "transformedSize": 0,
      },
      "version": "3.0.1",
    },
    {
      "id": 18,
      "modules": [
        44,
      ],
      "name": "color-name",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 4617,
        "transformedSize": 0,
      },
      "version": "1.1.4",
    },
    {
      "id": 19,
      "modules": [
        45,
        49,
        50,
      ],
      "name": "color-convert",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 20802,
        "transformedSize": 0,
      },
      "version": "1.9.3",
    },
    {
      "id": 20,
      "modules": [
        48,
        52,
      ],
      "name": "dom-serializer",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 10115,
        "transformedSize": 0,
      },
      "version": "1.4.1",
    },
    {
      "id": 21,
      "modules": [
        51,
        54,
        55,
        56,
        57,
        58,
        59,
        60,
      ],
      "name": "entities",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 46808,
        "transformedSize": 0,
      },
      "version": "2.2.0",
    },
    {
      "id": 22,
      "modules": [
        53,
      ],
      "name": "color-name",
      "root": "",
      "size": {
        "parsedSize": 0,
        "sourceSize": 4617,
        "transformedSize": 0,
      },
      "version": "1.1.3",
    },
  ],
}
`;
